File: wolkfind/deeptrail.py
# /// script
# dependencies = ["rich"]
# ///
from __future__ import annotations

import gzip
import json
import os
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional, Tuple

import boto3
from rich.console import Console


def get_trail_config(session: boto3.Session, console: Console, verbose: bool) -> Optional[Tuple[str, str]]:
    ct = session.client("cloudtrail", region_name="us-east-1")
    trails = ct.describe_trails().get("trailList", [])
    if not trails:
        return None
    trail = trails[0]
    if verbose:
        console.print(
            f"[dim blue]Using Trail: {trail['Name']} (S3: {trail['S3BucketName']})[/]"
        )
    return trail.get("S3BucketName"), trail.get("S3KeyPrefix", "")


def download_trail_samples(
    session: boto3.Session,
    account_id: str,
    account_dir: str,
    region: str,
    bucket: str,
    prefix: str,
    days: int,
    verbose: bool,
    console: Console,
    safe_api_call,
) -> List[str]:
    s3 = session.client("s3")
    now = datetime.now(timezone.utc)
    downloaded: List[str] = []
    cache_dir = os.path.join(account_dir, region, "cache", "cloudtrail")
    os.makedirs(cache_dir, exist_ok=True)

    for i in range(days):
        dt = now - timedelta(days=i)
        day_prefix = (
            f"{prefix}/AWSLogs/{account_id}/CloudTrail/{region}/"
            f"{dt.year}/{dt.month:02d}/{dt.day:02d}/"
        )
        if prefix == "":
            day_prefix = day_prefix[1:]

        objs = s3.list_objects_v2(Bucket=bucket, Prefix=day_prefix, MaxKeys=5).get(
            "Contents", []
        )
        if not objs:
            continue
        obj = objs[0]
        key = obj["Key"]
        fname = key.split("/")[-1]
        local_path = os.path.join(cache_dir, fname)
        if not os.path.exists(local_path):
            if verbose:
                console.print(
                    f"  [dim green]Downloading sample for {dt.strftime('%Y-%m-%d')}...[/]"
                )
            safe_api_call("s3", "download_file", s3.download_file, bucket, key, local_path)
        if os.path.exists(local_path):
            downloaded.append(local_path)
    return downloaded


def parse_trail_events(file_paths: List[str]) -> List[Dict]:
    events: List[Dict] = []
    for path in file_paths:
        try:
            with gzip.open(path, "rb") as f:
                data = json.load(f)
                events.extend(data.get("Records", []))
        except Exception:
            continue
    return events

File: wolkfind-go/pkg/discovery/deeptrail.go
package discovery

import (
	"compress/gzip"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudtrail"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/pterm/pterm"
)

type TrailConfig struct {
	Bucket string
	Prefix string
}

func GetTrailConfig(ctx context.Context, cfg aws.Config, verbose bool) (*TrailConfig, error) {
	// Need to query cloudtrail in us-east-1 (or current region)
	ctClient := cloudtrail.NewFromConfig(cfg)
	resp, err := ctClient.DescribeTrails(ctx, &cloudtrail.DescribeTrailsInput{})
	if err != nil || len(resp.TrailList) == 0 {
		return nil, fmt.Errorf("no trails found or failed to describe: %v", err)
	}

	trail := resp.TrailList[0]
	if verbose {
		pterm.Info.Printf("Using Trail: %s (S3: %s)\n", *trail.Name, *trail.S3BucketName)
	}

	prefix := ""
	if trail.S3KeyPrefix != nil {
		prefix = *trail.S3KeyPrefix
	}

	return &TrailConfig{
		Bucket: *trail.S3BucketName,
		Prefix: prefix,
	}, nil
}

func DownloadTrailSamples(ctx context.Context, cfg aws.Config, accountID, accountDir, region, bucket, prefix string, days int, verbose bool) ([]string, error) {
	s3Client := s3.NewFromConfig(cfg)
	var downloadedPaths []string
	now := time.Now().UTC()

	cacheDir := filepath.Join(accountDir, region, "cache", "cloudtrail")
	if err := os.MkdirAll(cacheDir, 0755); err != nil {
		return nil, err
	}

	for i := 0; i < days; i++ {
		dt := now.AddDate(0, 0, -i)
		dayPrefix := fmt.Sprintf("%s/AWSLogs/%s/CloudTrail/%s/%d/%02d/%02d/",
			prefix, accountID, region, dt.Year(), dt.Month(), dt.Day())
		
		// If prefix is empty, we might have double slashes
		dayPrefix = filepath.Clean(dayPrefix)
		if strings.HasPrefix(dayPrefix, "/") {
			dayPrefix = dayPrefix[1:]
		}

		resp, err := s3Client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket: aws.String(bucket),
			Prefix: aws.String(dayPrefix),
			MaxKeys: aws.Int32(5),
		})

		if err != nil || len(resp.Contents) == 0 {
			continue
		}

		obj := resp.Contents[0]
		key := *obj.Key
		fname := filepath.Base(key)
		localPath := filepath.Join(cacheDir, fname)

		if _, err := os.Stat(localPath); os.IsNotExist(err) {
			if verbose {
				pterm.Info.Printf("  Downloading sample for %s...\n", dt.Format("2006-01-02"))
			}
			
			getObjResp, err := s3Client.GetObject(ctx, &s3.GetObjectInput{
				Bucket: aws.String(bucket),
				Key:    aws.String(key),
			})
			if err != nil {
				continue
			}

			f, err := os.Create(localPath)
			if err != nil {
				getObjResp.Body.Close()
				continue
			}
			_, err = io.Copy(f, getObjResp.Body)
			f.Close()
			getObjResp.Body.Close()
			if err != nil {
				continue
			}
		}

		downloadedPaths = append(downloadedPaths, localPath)
	}

	return downloadedPaths, nil
}

func ParseTrailEvents(paths []string) ([]map[string]interface{}, error) {
	var events []map[string]interface{}
	for _, path := range paths {
		f, err := os.Open(path)
		if err != nil {
			continue
		}
		
		gz, err := gzip.NewReader(f)
		if err != nil {
			f.Close()
			continue
		}

		var data struct {
			Records []map[string]interface{} `json:"Records"`
		}
		if err := json.NewDecoder(gz).Decode(&data); err == nil {
			events = append(events, data.Records...)
		}
		
		gz.Close()
		f.Close()
	}
	return events, nil
}
